File: Data Structures & Algorithms/two-integer-sum/submission-1.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // we have to make a hash map for value : index
        unordered_map<int, int> prevMap; // value : index
        for (int i = 0; i < nums.size(); i++){
            int value = target - nums[i];
            if (prevMap.find(value) != prevMap.end()) return {prevMap[value], i};
            prevMap.insert({nums[i], i});
        }
        return {};
    }
};

File: Data Structures & Algorithms/find-duplicate-integer/submission-1.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow = nums[0]
        fast = nums[0]

        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break
        
        slow = nums[0]
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]
        
        return slow
"""
 Input: nums = [1,2,3,2,2]

    index = 0
    nums[index] = 1
    nums[1] = 2
    nums[2] = 3
    nums[3] = 2
    nums[2] = 3
    nums[3] = 2
              Y
         |---------|
  X      |         \\
---------|         |
         |---------|
         Z

slow = X + Y
fast = X + 2Y + Z

2(slow) = fast
2(X+Y) = X+2Y+Z
2X + 2Y = X + 2Y + Z
X = Z """